update 'unsafe' (#1278)
rust-lang/reference: src/unsafe-keyword.md

# The `unsafe` keyword

> The `unsafe` keyword in Rust can occur in several different context: unsafe

Reviewer: We try to avoid "in Rust" since everything in the reference is about Rust. I also think this seems like it should be plural.

```suggestion
The `unsafe` keyword can occur in several different contexts: unsafe
```

functions (`unsafe fn`), unsafe blocks (`unsafe {}`), unsafe traits (`unsafe
trait`), and unsafe trait implementations (`unsafe impl`). It plays several
different roles, depending on where it is used and whether the

> `unsafe_op_in_unsafe_fn` lint is enabled:

Reviewer: I'm always on the fence on referring to rustc's lints in the reference since they aren't exactly part of the language. However, in this case it seems very relevant here. I think we'll need to figure out a better policy in the future.

It might help to link this for someone to learn more about the lint:

```suggestion
[`unsafe_op_in_unsafe_fn`] lint is enabled:
```

```
[`unsafe_op_in_unsafe_fn`]: ../rustc/lints/listing/allowed-by-default.html#unsafe-op-in-unsafe-fn
```

Author: Oh I didn't know each lint has a page like that. :)

- it is used to mark code that *defines* extra safety conditions (`unsafe fn`,
  `unsafe trait`)
- it is used to mark code that needs to *satisfy* extra safety conditions
  (`unsafe {}`, `unsafe impl`, `unsafe fn` without `unsafe_op_in_unsafe_fn`)

The following discusses each of these cases. See the
[keyword documentation][keyword] for some illustrative examples.

[keyword]: ../std/keyword.unsafe.html

## Unsafe functions (`unsafe fn`)

Unsafe functions are functions that are not safe in all contexts and/or for all
possible inputs. We say they have *extra safety conditions*, which are
requirements that must be upheld by all callers and that the compiler does not

> check. For example, `get_unchecked` has the extra safety

Reviewer: I think it would be good to link this for context.

```suggestion
check. For example, [`get_unchecked`] has the extra safety
```

```
[`get_unchecked`]: ../std/primitive.slice.html#method.get_unchecked
```

condition that the index must be in-bounds. The module defining an unsafe

> function is responsible for documenting what those extra safety conditions are.

Reviewer: This seems rather specific to say it must go in the *module* documentation. For example, the aforementioned `get_unchecked` is just documented on the method itself, not in the module docs. Can this be reworded to be a little more general?

Reviewer 2: I'm going to say that we should point out that this isn't an actual requirement of the language, but rather expected practice. There's nothing in `rustc` that will (or could) stop your code from compiling because you failed to explain what extra conditions you are imposing.

Author: I didn't mean this to imply it must be the module-level docs, just 'somewhere inside the module'.

Such a function must be prefixed with the keyword `unsafe` and can only be

> called from inside an `unsafe` block.

Reviewer: ```suggestion
called from inside an `unsafe` function or block.
```

unsafe function is also considered to be an unsafe block; this can be changed by
enabling the `unsafe_op_in_unsafe_fn` lint.

By putting operations into an unsafe block, the programmer states that they have
taken care of satisfying the extra safety conditions of all operations inside
that block.

Unsafe blocks are the logical dual to unsafe functions: where unsafe functions
define a proof obligation that callers must uphold, unsafe blocks state that all
relevant proof obligations have been discharged. There are many ways to
discharge proof obligations; for example, there could be run-time checks or data
structure invariants that guarantee that certain properties are definitely true,
or the unsafe block could be inside an `unsafe fn` and use its own proof
obligations to discharge the proof obligations of its callees.

Unsafe blocks are used to wrap foreign libraries, make direct use of hardware
or implement features not directly present in the language. For example, Rust
provides the language features necessary to implement memory-safe concurrency
in the language but the implementation of threads and message passing in the
standard library uses unsafe blocks.

> Rust's type system is a conservative approximation of the dynamic safety

Reviewer 2: Small wording quibble. Is it an approximation, or is it a subset?

(You can also just change "Rust's" to "The"

Author: "conservative approximation" means subset or superset, depending on which direction is the safe one.

I didn't write this part, but I think I prefer "conservative approximation" over subset or superset, since with subset/superset it is often ambiguous what the polarity of the order is.

requirements, so in some cases there is a performance cost to using safe code.
For example, a doubly-linked list is not a tree structure and can only be
represented with reference-counted pointers in safe code. By using `unsafe`
blocks to represent the reverse links as raw pointers, it can be implemented

> with only boxes.

Reviewer 2: I have a feeling you are purposely hinting at https://rust-unofficial.github.io/too-many-lists/

I wonder if we could explicitly just tell people they can read that book.

Author: This is old text, I didn't write this ;)

## Unsafe traits (`unsafe trait`)

An unsafe trait is a trait that comes with extra safety conditions that must be
upheld by *implementations* of the trait. The module defining an unsafe trait is

> responsible for documenting what those extra safety conditions are.

Reviewer: Ditto about being specific about the docs being in the module.

Author: I have adjusted the wording, please let me know what you think.

Such a trait must be prefixed with the keyword `unsafe` and can only be
implemented by `unsafe impl` blocks.

## Unsafe trait implementations (`unsafe impl`)

When implementing an unsafe trait, the implementation needs to be prefixed with
the `unsafe` keyword. By writing `unsafe impl`, the programmer states that they
have taken care of satisfying the extra safety conditions required by the trait.

> Unsafe traits are the logical dual to unsafe traits: where unsafe traits define

Reviewer: ???

```suggestion
Unsafe traits are the logical dual to unsafe implementations: where unsafe traits define
```

Author: :joy:
